File: 파이썬 알고리즘 인터뷰/12장 그래프/조합.js
// 나의 풀이 (Runtime: 152ms, Memory: 46.7MB)
const combine = function(n, k) {
    const result = [];
    const prevNumbers = [];
    const numbers = [];
    for (let i = 1; i <= n; i++) {
        numbers.push(i);
    }
    
    const combination = function(numbers, count) {
        // 주어진 조합의 길이를 달성한 경우, 함수를 종료한다.
        if (count === 0) {
            result.push(prevNumbers.slice());
            return;
        }
        
        numbers.forEach((number, idx, fullArr) => {
            // 자신과 자신 이전의 숫자를 제외한 배열을 생성한다.
            const nextNumbers = fullArr.slice(idx + 1);
            
            prevNumbers.push(number);
            combination(nextNumbers, count - 1);
            prevNumbers.pop(); // 백트래킹
        })
    }
    
    combination(numbers, k);
    
    return result;
};

File: 파이썬 알고리즘 인터뷰/12장 그래프/조합.py
class Solution:
    # 직접 구현한 풀이 (Runtime: 440ms, Memory: 15.8MB)
    def combine(self, n: int, k: int) -> List[List[int]]:
        result = []
    
        def dfs(numbers, start, k):
            if k == 0:
                result.append(numbers[:])
                return
            
            # 자기 자신 이후의 모든 수에 대해 재귀 호출.
            for idx in range(start, n + 1):
                numbers.append(idx)
                dfs(numbers, idx + 1, k - 1)
                numbers.pop()  # 백트래킹
        
        dfs([], 1, k)
        
        return result
    
    # itertools를 사용한 풀이 (Runtime: 68ms, Memory: 15MB)
    def combine(self, n: int, k: int) -> List[List[int]]:
        return itertools.combinations([i for i in range(1, n + 1)], k)